---
title: "Structural Conservation"
author: "Michael Soutschek"
date: "4/17/2021"
output: html_document
---



```{r}
suppressPackageStartupMessages({
  library(SummarizedExperiment)
  library(tidyverse)
  library(ggpubr)
  library(GenomicRanges)
  library(msa)
})
```

```{r}
readRNAz <- function(file_name) {
  file <- read_file(file_name)
  entries <- strsplit(file, ">>")[[1]]  
  rnaz_parsed <- lapply(entries, parseRNAz_entry)
  df <- do.call(rbind, rnaz_parsed)
  numeric_cols <- c(2:3,5:14,17:18)
  for(i in numeric_cols) df[[i]] <- as.numeric(df[[i]])
  return(df)
}

parseRNAz_entry <- function(entry) {
  entry_list <- strsplit(entry, "\n")[[1]]
  if(length(entry_list)<2) return(data.frame())
  miRNA <- entry_list[[1]]
  trimmed <- lapply(5:22, function(i) trimws(strsplit(entry_list[[i]], ":")[[1]]))
  df <- as.data.frame(do.call(cbind, lapply(trimmed, `[[`, 2)))
  names(df) <- gsub(" ", "\\.", lapply(trimmed, `[[`, 1))
  return(cbind(miRNA = miRNA, df))
}
```



#Prepare data
```{r}
## load RNAz results for primates
df_rnaz <- readRNAz("../data/rnaz/rnaz_results_all.txt")


df_rnaz_primates <- readRNAz("../data/rnaz/rnaz_results_primates.txt")
df_seqid <- readRDS("../res/seq_identity.rds")

## pairwise rnaz for all organisms
non_humans <- df_seqid %>% 
  dplyr::select(-miRNA) %>%
  names()



df_rnaz_pairwise <- readRDS("../res/rnaz_results_pairwise.rds")

# filter out miRNAs that have an identical ortholog somewhere
id_miRNAs <- df_seqid %>%
  dplyr::filter_at(vars(-miRNA), any_vars(.==100)) %>%
  pull(miRNA)

# get highest structure conservation for each miRNA
df_rnaz_pairwise_filtered <- df_rnaz_pairwise %>%
  dplyr::filter(!miRNA %in% id_miRNAs) %>%
  group_by(miRNA) %>%
  dplyr::slice(which.max(Structure.conservation.index))
```



```{r}
## load expression data
se_smallRNA <- readRDS("../../../TimeCourse_GRCh38.p10_Ens91/data/Small_RNA/smallRNA_oasis.DEA.SE.rds")

## associate expression of miRNAs with precursors
grl <- rtracklayer::import("../data/mirnas/hsa.gff3")

smallRNA_counts <- se_smallRNA  %>%
  assays() %>%
  `[[`("logcpm") %>%
  as.data.frame() %>%
  rownames_to_column(var = "ID") %>%
  pivot_longer(cols = starts_with("SC")) %>%
  group_by(ID) %>%
  summarize(
    min_exp = min(value),
    max_exp = max(value),
    mean_exp = mean(value),
    median_exp = median(value)) %>%
  inner_join(mcols(grl)[c("Name", "Derives_from")], by = c("ID" = "Name"), copy=TRUE) %>%
  left_join(mcols(grl)[c("Name", "ID")], by = c("Derives_from" = "ID"), copy=TRUE) %>%
  dplyr::select(-Derives_from) %>%
  dplyr::rename(precursor = Name)


## integrate with expression data
df_rnaz_primates_exp <- df_rnaz_primates %>%
  left_join(smallRNA_counts, by = c("miRNA" = "precursor")) %>%
  mutate(expressed = !is.na(min_exp))

df_rnaz_pairwise_exp <- df_rnaz_pairwise_filtered %>%
  left_join(smallRNA_counts, by = c("miRNA" = "precursor")) %>%
  mutate(expressed = !is.na(min_exp))
```


```{r}
## get precursor names of miRNAs with seed mismatches
miRNA_alignments <- readRDS("../res/miRNA_alignments_seed.rds")
mature_seed_mismatch <- names(miRNA_alignments)
precursor_IDs <- grl[grl$Name %in% mature_seed_mismatch]$Derives_from
miRNAs_seed_mismatch <- grl[grl$ID %in% precursor_IDs]$Name

df_plot_primates <- df_rnaz_primates_exp %>%
  dplyr::filter(expressed) %>%
  mutate(seed_mismatch = miRNA %in% miRNAs_seed_mismatch) %>%
  distinct(miRNA,.keep_all = TRUE)



df_plot_pairwise <- df_rnaz_pairwise_exp %>%
  dplyr::filter(expressed) %>%
  mutate(seed_mismatch = miRNA %in% miRNAs_seed_mismatch) %>%
  distinct(miRNA,.keep_all = TRUE)
```




#Plot
```{r}
#show_col(inferno(10))
yellow <- "#FB9A06"
orange <- "#ED6925"
violet <- "#4B0C6B"
colors <- c(violet,orange)

p_primates <- ggplot(df_plot_primates, aes(x = Mean.pairwise.identity, 
                                           y = Structure.conservation.index,
                                           color = seed_mismatch)) +
  geom_smooth(method = "lm", se = FALSE, linetype = "dashed", color = "darkgrey") +
  geom_abline(slope = 0.01,intercept = 0, linetype = "dashed", color = "black") +
  geom_point() +
  geom_point(data = df_plot_primates[df_plot_primates$seed_mismatch == "TRUE",], shape = 1, size = 4) +
  scale_color_manual(values = colors) +
  ggrepel::geom_text_repel(aes(label = ifelse(Structure.conservation.index < 0.9 | seed_mismatch == "TRUE", miRNA, "")), size = 3, show.legend = F, max.overlaps = Inf) +
  ggtitle("Structural conservation among all Primates") +
  xlab("Mean pairwise identity")+
  ylab("Structure conservation index")+
  labs(color = "Seed Mutation") + 
  theme_classic(base_size = 12) +
  theme(legend.position = "bottom")

p_pairwise <- ggplot(df_plot_pairwise, aes(x = Mean.pairwise.identity, 
                    y = Structure.conservation.index)) +
  geom_smooth(method = "lm", se = FALSE, linetype = "dashed", color = "darkgrey") +
  geom_abline(slope = 0.01,intercept = 0, linetype = "dashed", color = "black") +
  geom_point(aes(color = seed_mismatch)) +
  geom_point(data = df_plot_pairwise[df_plot_pairwise$seed_mismatch == "TRUE",], aes(color = seed_mismatch), shape = 1, size = 4) +
  scale_color_manual(values = colors) +
  ggrepel::geom_text_repel(aes(label = ifelse(Structure.conservation.index < 0.975 | seed_mismatch == "TRUE", miRNA, "")), size = 3, show.legend = F, max.overlaps = 10, point.padding = 0.5) +
  ggtitle("Structural conservation pairwise versus human") +
  xlab("Mean pairwise identity")+
  ylab("Max. structure conservation index")+
  labs(color = "Seed Mutation") + 
  coord_cartesian(ylim = c(0.85,NA)) +
  theme_classic(base_size = 12) +
  theme(legend.position = "bottom")


p <- ggarrange(p_primates, p_pairwise, ncol = 1)

ggsave("./Figure_Output/structural_conservation.png",p, width = 7, height = 8)


p_pairwise2 <- ggplot(df_plot_pairwise, aes(x = Mean.pairwise.identity, 
                    y = Structure.conservation.index)) +
  geom_smooth(method = "lm", se = FALSE, linetype = "dashed", color = "darkgrey") +
  geom_abline(slope = 0.01,intercept = 0, linetype = "dashed", color = "black") +
  geom_point(aes(color = seed_mismatch)) +
  geom_point(data = df_plot_pairwise[df_plot_pairwise$seed_mismatch == "TRUE",], aes(color = seed_mismatch), shape = 1, size = 4) +
  scale_color_manual(values = colors) +
  ggrepel::geom_text_repel(aes(label = ifelse(Structure.conservation.index < 0.975 | seed_mismatch == "TRUE", miRNA, "")), size = 3, show.legend = F, max.overlaps = 10, point.padding = 0.5) +
  ggtitle("Structural conservation pairwise versus human") +
  xlab("Mean pairwise identity")+
  ylab("Max. structure conservation index")+
  labs(color = "Seed Mutation") + 
  coord_cartesian(ylim = c(0.85,NA)) +
  theme_classic(base_size = 12) +
  theme(legend.position = "bottom") +
  theme(text = element_text(size = 14), 
        axis.text.x = element_text(size = rel(1))
        )


ggsave("./Figure_Output/structural_conservation_pairwise.png",p_pairwise2, width = 7, height = 4.5)
```






