---
title: "ms_proteomics_all-tp_DEG"
author: "Michael"
date: "08 01 2020"
output: 
    html_document:
        toc: true
        toc_float: true
        theme: 'cerulean'
        highlight: 'tango'
        code_folding: hide
        df_print: paged
---

<style>
  .main-container {
    max-width: 1500px !important;
  }
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
suppressPackageStartupMessages({
  library(SummarizedExperiment)
  library(SEtools)
  library(ggplot2)
  library(plgINS)
  library(viridis)
  library(RColorBrewer)
  library(data.table)
  library(dplyr)
  library(tidyr)
  library(DEP)
  library(limma)
})
```


Load data

```{r}
data <- read.table("../data/Proteomics/20191210_130249_20191209_iPSCScreen_Report_Protein.csv", sep = ",", header = T, na.strings = "NaN")
s2c <- read.table("../data/Proteomics/metadata.csv", sep =";", header = T)
s2c$Day <- as.character(s2c$Day)
assign <- match(data$R.FileName,s2c$filename)
data$Condition <- s2c$Condition[assign]
data$Experiment <- s2c$Experiment[assign]
data$Celltype <- s2c$CellType[assign]
data$Day <- s2c$Day[assign]
data$expReplicate <- s2c$expReplicate[assign]
```

Opionally, to also load Peptide Data

```{r}
# data_pep <- readRDS("../data/Proteomics/20191210_iPSCScreen_Report_Peptide_Table.rds")
```

Create a Matrix of Expression Data

```{r}
matr <- spread(data[,c("R.FileName","PG.ProteinAccessions","PG.Quantity")], key = R.FileName, value = PG.Quantity)
row.names(matr) <- matr[,1]
matr[,1] <- NULL
keep <- apply(matr, 1, function(x) sum(!is.na(x))> 0)
matr <- matr[keep,]

# saveRDS(matr, file = "../data/Proteomics/20191210_iPSCScreen_Report_Protein_Expression_Matrix.rds")
```

```{r}
# matr_pep <- spread(data_pep[,c("R.FileName","EG.PrecursorId","FG.Quantity")], key = R.FileName, value = FG.Quantity)
# row.names(matr_pep) <- matr_pep[,1]
# matr_pep[,1] <- NULL
# keep_pep <- apply(matr_pep, 1, function(x) sum(!is.na(x))> 0)
# matr_pep <- matr_pep[keep_pep,]
# 
# saveRDS(matr_pep, file = "../data/Proteomics/20191210_iPSCScreen_Report_Peptide_Expression_Matrix.rds")
```



DEP, Prepare SE and Impute
```{r}
set.seed(1234)
matr$Protein.names <- row.names(matr)
matr$Protein.IDs <- paste("ID_",c(1:nrow(matr)),sep = "")
data_unique <- make_unique(matr, "Protein.names", "Protein.IDs", delim = ";")

experimental_design <- data.frame(label = as.character(s2c[,"filename"]), 
                                  condition = as.character(s2c[,"Condition"]), 
                                  replicate = as.integer(s2c[,"expReplicate"]))
experimental_design[,c("label","condition")] <- lapply(experimental_design[,c("label","condition")], as.character)

LFQ_columns <- na.omit(match(s2c$filename,colnames(data_unique)))
data_se <- make_se(data_unique, LFQ_columns, experimental_design)
data_filt <- filter_missval(data_se, thr = 0)
assayNames(data_filt) <- "intensity"
data_norm <- normalize_vsn(data_filt)
assays(data_filt)$vsn <- assay(data_norm)
#normalization is not absolutely necessary because we normalize already with the spectronaut
se_imp <- impute(data_norm, fun = "MinProb", q = 0.01)
assays(data_filt)$imputed <- assay(se_imp)
se_prot <- data_filt
se_prot <- se_prot[,order(se_prot$condition, se_prot$replicate)]

#get the "day" info
se_prot$condition[1:3] <- paste0(se_prot$condition[1:3],"_0")
se_prot$day <- sapply(strsplit(se_prot$condition,"_"),"[",2)
se_prot$day[4:6] <- gsub("0","",se_prot$day[4:6])
se_prot$day <- as.numeric(se_prot$day)
se_prot$stage <- as.factor(se_prot$day)

#add further info
assign_info <- match(names(se_prot), data$PG.ProteinAccessions)
rowData(se_prot)$StrippedSequences <- data[assign_info, "PG.NrOfStrippedSequencesMeasured"]
rowData(se_prot)$GeneName <- data[assign_info,"PG.Genes"]
rowData(se_prot)$UniProtName <- data[assign_info, "PG.ProteinNames"]
rowData(se_prot)$ProteinName <- data[assign_info, "PG.ProteinAccessions"]

row.names(se_prot) <- paste0(row.names(se_prot),".",rowData(se_prot)$GeneName)

#Safe the raw SE-object
saveRDS(se_prot, file="../data/Proteomics/Proteomics.SE_raw.rds")
saveRDS(data,file="../data/Proteomics/data_raw.rds")
```



#################################################################
# Statistical Analysis
#################################################################

#load data
```{r}
se <- readRDS("../data/Proteomics/Proteomics.SE_raw.rds")
data <- readRDS("../data/Proteomics/data_raw.rds")
data$ID <- paste0(data$PG.ProteinAccessions,".",data$PG.Genes)
```

#LogFC
```{r}
assays(se)$log2FC.all <-
  assays(se)$imputed-rowMeans(assays(se)$imputed[,which(se$day=="0")])

assays(se)$log2FC.neur <-
  assays(se)$imputed-rowMeans(assays(se)$imputed[,which(se$day=="9")])
```


#Proteomics DEA function
```{r}
prot.dea <- function(se,data,use = NULL,name,model){
  #filter
  if(!is.null(use) ){
    se.sub <- se[,use]
  }else{
    se.sub <- se
  }
  colData(se.sub) <- droplevels(colData(se.sub))
  
  #DEA
  dfp <- as.data.frame(assays(se.sub)$imputed)
  mmp <- model.matrix(model, data=as.data.frame(colData(se.sub)))
  fitp <- eBayes(lmFit(dfp, mmp))
  resp <- as.data.frame(topTable(fitp, number=Inf))
  resp$ID <- row.names(resp)
  assign2 <- match(resp$ID, data$ID)
  resp$StrippedSequences <- data[assign2,"PG.NrOfStrippedSequencesMeasured"]
  resp$GeneName <- data[assign2,"PG.Genes"]
  resp$UniProtName <- data[assign2,"PG.ProteinNames"]
  resp <- resp[rownames(se.sub),]
  
   # add DEA results to SE
  dea_name <- paste0("DEA.", name)
  rowData(se)[[dea_name]] <- DataFrame(resp)
  
  return(se)
}
```


## dea over all days
```{r}
se <- prot.dea(se = se,data = data, name = "all", model = ~stage)
```

## dea over neuronal days
```{r}
se <- prot.dea(se = se, data = data, use = which(se$day > 0), name = "neuronal", model = ~stage)
```


## pairwise DEA over all stages

```{r variables, eval=TRUE}
# define variables
stages <- unique(as.numeric(as.character(se$stage)))
dea.names <- c("9v0", "15v9", "21v15", "27v21", "33v27", "40v33")
```

```{r dea pairwise, eval=TRUE}
# do pairwise differential expression analysis for all stages
for(i in 1:length(dea.names)){
  se <- prot.dea(se = se, data = data, name = dea.names[i], 
                 use = se$stage %in% c(stages[i],stages[i+1]), model = ~stage)
}
```

Safe the object
```{r}
saveRDS(se, file="../data/Proteomics/Proteomics.SE_DEA.rds")
```


